doc: Improve `dependencies.md`

NicolaLS opened the pull request:
Small improvements to the `dependencies.md` documentation as a follow-up for #31634.

**Linux Kernel** does not need to be in the dependencies as it is not required for cross-compiling from other systems, and users building on Linux should not expect they can build using any EOL kernel, see: https://github.com/bitcoin/bitcoin/pull/31634#discussion_r1957123270

**Runtime dependencies** can be in a separate table to improve readability. See: https://github.com/bitcoin/bitcoin/pull/31634#issuecomment-2589412550

**Version used** is redundant as the depends package definition is already linked in the table and can thus be removed, see: https://github.com/bitcoin/bitcoin/pull/31895#discussion_r2063356972

NicolaLS committed:
doc: Remove Linux Kernel from dep. table

Linux Kernel is neither an optional nor a required dependency for
Bitcoin Core.

NicolaLS committed:
doc: Improve dependencies.md documentation

- Remove the "Version used" column from the dependencies tables as the
  depends package definition which defines the version used is already
  linked. In case a developer is interested in which PR introduced this
  file/version they can use `git blame` on the package definition as
  usual. This removes doc. maintenance overhead and eliminates the risk
  of stale information about the "Version used", see comment:
  https://github.com/bitcoin/bitcoin/pull/31895#discussion_r2063356972

- Separate dependency tables into build-time and run-time tables for
  easier distinction of the two and to avoid repeating the same
  information ("No"/"Yes") for better readability.

- Order dependencies alphabetically

NicolaLS committed:
doc: Add dependency self-compilation info

The `dependencies.md` should mention that it is possible to self-compile
the dependencies and reference `depends/README.md` for instructions.
Also mention full path to `/doc/build-*.md` for clarity.

NicolaLS commented:
[Preview](https://github.com/NicolaLS/bitcoin/blob/doc-followup/doc/dependencies.md)

hodlinator commented on doc/dependencies.md:
```diff
@@ -15,24 +15,37 @@ Bitcoin Core requires one of the following compilers.
+
+| Dependency | Releases | Version used | Minimum required |
+| --- | --- | --- | --- |
+| [qrencode](../depends/packages/qrencode.mk) (gui) | [link](https://fukuchi.org/works/qrencode/) | [4.1.1](https://github.com/bitcoin/bitcoin/pull/27312) | N/A |
+| [Qt](../depends/packages/qt.mk) (gui) | [link](https://download.qt.io/official_releases/qt/) | [5.15.16](https://github.com/bitcoin/bitcoin/pull/30774) | [5.11.3](https://github.com/bitcoin/bitcoin/pull/24132) |
+| [ZeroMQ](../depends/packages/zeromq.mk) (notifications) | [link](https://github.com/zeromq/libzmq/releases) | [4.3.4](https://github.com/bitcoin/bitcoin/pull/23956) | 4.0.0 |
+| [Berkeley DB](../depends/packages/bdb.mk) (legacy wallet) | [link](https://www.oracle.com/technetwork/database/database-technologies/berkeleydb/downloads/index.html) | 4.8.30 | 4.8.x |
+| [SQLite](../depends/packages/sqlite.mk) (wallet) | [link](https://sqlite.org) | [3.38.5](https://github.com/bitcoin/bitcoin/pull/25378) | [3.7.17](https://github.com/bitcoin/bitcoin/pull/19077) |
+| Python (scripts, tests) | [link](https://www.python.org) | [3.10](https://github.com/bitcoin/bitcoin/pull/30527) | [3.10](https://github.com/bitcoin/bitcoin/pull/30527) |
+| [systemtap](../depends/packages/systemtap.mk) ([tracing](tracing.md)) | [link](https://sourceware.org/systemtap/) | [4.8](https://github.com/bitcoin/bitcoin/pull/26945)| N/A |
```
Might be so cheeky as to order these alphabetically? Also CMake/Boost at the top.

NicolaLS replied:
Sure :)

NicolaLS replied:
14c85a9b3753b40e8cf9f2808530fa922c8a3473

hodlinator replied:
```
₿ touch A
₿ touch a
₿ ls [Aa]
a  A
```
Doesn't match GNU `ls` order.. but it does match ASCII order. :+1:

hodlinator replied:
Sorry I lost track of this PR. Will try to get to reviewing after rebase, maybe not this week though.

NicolaLS replied:
no worries, thanks :) (rebased now)

hodlinator reviewed (approved):
ACK f3d2e9acbf3d9407101290ca86f014e82344cfce

#### Concept

Separates build and runtime dependencies into separate tables which makes more sense.

#### Nits

* Slight preference of using `### Build` over `**Build:**`, but it's subjective.

* Maybe 32bdc7dc47bcd67458bc8f687fcf8659ec3f0b1e could also include a link to https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/cmake.scm?id=53396a22afc04536ddf75d8f82ad2eafa5082725#n169 in the commit message, since it's a bit opaque otherwise?

NicolaLS commented:
> ACK [f3d2e9a](https://github.com/bitcoin/bitcoin/commit/f3d2e9acbf3d9407101290ca86f014e82344cfce)
> #### Concept
> 
> Separates build and runtime dependencies into separate tables which makes more sense.
> #### Nits
> 
>     * Slight preference of using `### Build` over `**Build:**`, but it's subjective.
> 
>     * Maybe [32bdc7d](https://github.com/bitcoin/bitcoin/commit/32bdc7dc47bcd67458bc8f687fcf8659ec3f0b1e) could also include a link to https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/cmake.scm?id=53396a22afc04536ddf75d8f82ad2eafa5082725#n169 in the commit message, since it's a bit opaque otherwise?

Thanks. Fixed both nits, I also think headlines are better.

hodlinator reviewed (approved):
re-ACK 8da43955376bde65fa517af462fd2a1514764d2b

PR addresses feedback from at least 3 long-time contributors (see comment links in [PR description](https://github.com/bitcoin/bitcoin/pull/31895#issue-2860463438)).

Changes: Addressed nits from [first ACK](https://github.com/bitcoin/bitcoin/pull/31895#pullrequestreview-2746984398).

Nice bonus that one can now link directly to a specific sub-heading.
https://github.com/bitcoin/bitcoin/blob/8da43955376bde65fa517af462fd2a1514764d2b/doc/dependencies.md#build-1

fanquake commented on doc/dependencies.md:
```diff
@@ -15,24 +15,37 @@ Bitcoin Core requires one of the following compilers.
+| Dependency | Releases | Version used | Minimum required |
+| --- | --- | --- | --- |
+| [Berkeley DB](../depends/packages/bdb.mk) (legacy wallet) | [link](https://www.oracle.com/technetwork/database/database-technologies/berkeleydb/downloads/index.html) | 4.8.30 | 4.8.x |
+| Python (scripts, tests) | [link](https://www.python.org) | [3.10](https://github.com/bitcoin/bitcoin/pull/30527) | [3.10](https://github.com/bitcoin/bitcoin/pull/30527) |
```
Yep. This seems to be morphing from general documentation, to, "what happens to be used in Guix", which isn't necessarily relevant. Same for CMake, I don't think version used should be listed here.

NicolaLS replied:
Agreed. Removed both commits (adding CMake version used/adding Python version used) and updated the PR description.

jonatack commented on doc/dependencies.md:
```diff
@@ -19,8 +19,8 @@ Bitcoin Core requires one of the following compilers.
 | --- | --- | --- | --- |
-| CMake | [link](https://cmake.org/) | N/A | [3.22](https://github.com/bitcoin/bitcoin/pull/30454) |
 | [Boost](../depends/packages/boost.mk) | [link](https://www.boost.org/users/download/) | [1.81.0](https://github.com/bitcoin/bitcoin/pull/26557) | [1.73.0](https://github.com/bitcoin/bitcoin/pull/29066) |
+| CMake | [link](https://cmake.org/) | [3.24.2](https://github.com/bitcoin/bitcoin/pull/30730) | [3.22](https://github.com/bitcoin/bitcoin/pull/30454) |
```
In commit fefe8fc3c38701bdc3a3b1c6e4d0c8885f94e75f ISTM the CMake entry shouldn't change in this commit, apart from ordering alphabetically (why does it change here)

NicolaLS replied:
ah yes messed up the rebase..fixed now.

jonatack commented on doc/dependencies.md:
```diff
@@ -1,8 +1,8 @@
 These are the dependencies used by Bitcoin Core.
-You can find installation instructions in the `build-*.md` file for your platform.
-"Runtime" and "Version Used" are both in reference to the release binaries.
+You can find installation instructions in the `build-*.md` file for your platform, or self-compile
```
In commit 0a6547d26b570625922478e7e993ad1c97ec3ab4 maybe reference from root (feel free to ignore)

```suggestion
You can find installation instructions in the `/doc/build-*.md` file for your platform, or self-compile
```

jonatack reviewed (commented):
Swung by here a couple times previously and this pull is looking better now.

ACK fefe8fc3c38701bdc3a3b1c6e4d0c8885f94e75f modulo question below

jonatack commented:
ACK 4b6171982a20d736b2c627ab9b7ea788b06af457

hodlinator reviewed (approved):
re-ACK 4b6171982a20d736b2c627ab9b7ea788b06af457

Changes since [previous ACK](https://github.com/bitcoin/bitcoin/pull/31895#pullrequestreview-2752731041): Addressed feedback (https://github.com/bitcoin/bitcoin/pull/31895#discussion_r2035011528 https://github.com/bitcoin/bitcoin/pull/31895#discussion_r2051045020).

fanquake commented on doc/dependencies.md:
```diff
@@ -15,24 +15,37 @@ Bitcoin Core requires one of the following compilers.
+| [qrencode](../depends/packages/qrencode.mk) (gui) | [link](https://fukuchi.org/works/qrencode/) | [4.1.1](https://github.com/bitcoin/bitcoin/pull/27312) | N/A |
+| [SQLite](../depends/packages/sqlite.mk) (wallet) | [link](https://sqlite.org) | [3.38.5](https://github.com/bitcoin/bitcoin/pull/25378) | [3.7.17](https://github.com/bitcoin/bitcoin/pull/19077) |
+| [systemtap](../depends/packages/systemtap.mk) ([tracing](tracing.md)) | [link](https://sourceware.org/systemtap/) | [4.8](https://github.com/bitcoin/bitcoin/pull/26945)| N/A |
+| [ZeroMQ](../depends/packages/zeromq.mk) (notifications) | [link](https://github.com/zeromq/libzmq/releases) | [4.3.4](https://github.com/bitcoin/bitcoin/pull/23956) | 4.0.0 |
```
Can you update zmq version used to 4.3.5 (https://github.com/bitcoin/bitcoin/pull/28627).

NicolaLS replied:
> Is there any value in documenting the version used twice with the additional maintenance overhead and risk of it being stale? The depends package definition is already linked, so anyone can get the real version with just one more click/open. If they want to find out the pull request/commit it will be just another step (git blame).
> 
> Could consider removing it?

I absolutely agree, I can add a commit to remove the "Version used" column from all tables if everyone is fine with this.

jonatack replied:
I don't have a strong opinion on removing the versions here; either way can probably have just one commit, e.g. "update dependency versions" or "remove versions"

NicolaLS replied:
Removed the sqlite/zmq update commits in favor of new: cb623692a6089c39560a637a2bf064d54aebb4d4

See https://github.com/bitcoin/bitcoin/pull/31895#issuecomment-2859720907

hodlinator replied:
nit: Feels weird to remove "Version used" column for runtime dependencies as part of the second commit (splitting build/runtime dependencies), but keeping them for buildtime dependencies until the end.

hodlinator replied:
I'd keep the removal of the Linux Kernel and addition of self-compilation info as separate commits and merge the other 3 into 1 - leaving 3 remaining commits. But squashing all is fine too.

fanquake commented on doc/dependencies.md:
```diff
@@ -15,24 +15,37 @@ Bitcoin Core requires one of the following compilers.
+| Python (scripts, tests) | [link](https://www.python.org) | N/A | [3.10](https://github.com/bitcoin/bitcoin/pull/30527) |
+| [Qt](../depends/packages/qt.mk) (gui) | [link](https://download.qt.io/archive/qt/) | [6.7.3](https://github.com/bitcoin/bitcoin/pull/30997) | [6.2](https://github.com/bitcoin/bitcoin/pull/30997) |
+| [qrencode](../depends/packages/qrencode.mk) (gui) | [link](https://fukuchi.org/works/qrencode/) | [4.1.1](https://github.com/bitcoin/bitcoin/pull/27312) | N/A |
+| [SQLite](../depends/packages/sqlite.mk) (wallet) | [link](https://sqlite.org) | [3.38.5](https://github.com/bitcoin/bitcoin/pull/25378) | [3.7.17](https://github.com/bitcoin/bitcoin/pull/19077) |
```
Can you update sqlite version used to 3.46.1 (#29991).

NicolaLS commented:
Rebased and dropped the two commits that bumped sqlite/zmq in favor of removing the "Version used" column from the tables entirely (see cb623692a6089c39560a637a2bf064d54aebb4d4).

This commit also removed _"Version Used" refers to the release binaries._ not sure if this should get replaced with something else ? (not necessary imo.)

hodlinator reviewed (approved):
re-ACK cb623692a6089c39560a637a2bf064d54aebb4d4

Changes since [previous ACK](https://github.com/bitcoin/bitcoin/pull/31895#pullrequestreview-2796426546):
* Rebased after BDB dependency removal.
* Removed remaining "Version used" columns based off https://github.com/bitcoin/bitcoin/pull/31895#discussion_r2063356972.

NicolaLS commented:
Squashed and reworded commits.

NicolaLS commented:
(fixed typo in commit message)

hebasto commented on doc/dependencies.md:
```diff
@@ -15,23 +15,36 @@ Bitcoin Core requires one of the following compilers.
+| Dependency | Releases | Minimum required |
+| --- | --- | --- |
+| [Boost](../depends/packages/boost.mk) | [link](https://www.boost.org/users/download/) | [1.73.0](https://github.com/bitcoin/bitcoin/pull/29066) |
+| CMake | [link](https://cmake.org/) | [3.22](https://github.com/bitcoin/bitcoin/pull/30454) |
```
nit: 
```suggestion
| CMake | [link](https://cmake.org/download/) | [3.22](https://github.com/bitcoin/bitcoin/pull/30454) |
```

hebasto replied:
> Anything here seems fine, we don't link to https://sqlite.org/download.html or https://freetype.org/download.html.

But we do link to https://www.boost.org/users/download/ and https://github.com/libevent/libevent/releases.

I’ve ACKed the current branch and am content with it as is.

hebasto commented on doc/dependencies.md:
```diff
@@ -15,23 +15,36 @@ Bitcoin Core requires one of the following compilers.
+
+| Dependency | Releases | Minimum required |
+| --- | --- | --- |
+| [Fontconfig](../depends/packages/fontconfig.mk) (gui) | [link](https://www.freedesktop.org/wiki/Software/fontconfig/) | 2.6 |
+| [FreeType](../depends/packages/freetype.mk) (gui) | [link](https://freetype.org) | 2.3.0 |
```
nit: From https://github.com/bitcoin/bitcoin/pull/30997#issue-2554496391:
> Starting with Qt 6.5.0, the `libxcb-cursor0` package is required to be installed at runtime.

hebasto replied:
It was a nit. I am OK with the current branch.

hebasto reviewed (approved):
ACK e62423d6f1514b022155edb5bc930cecc4236731.

jonatack reviewed (commented):
ACK e62423d6f1514b022155edb5bc930cecc4236731
